### ISSUE TITLE
[HOLD for payment 2024-07-24] [$250] Room - Room mention is not copied to clipboard and editor is empty when editing room mention

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#40403]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.63-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition:
- User has created a few rooms linked to the same workspace.

1. Go to staging.new.expensify.com
2. Go to any room from the precondition.
3. Type # and select a room.
4. Send the message.
5. Right click on the message > Copy to clipboard.
6. Paste the content.
7. Right click on the message > Edit comment.

## Expected Result:
In Step 5, the room in the message will be copied.
In Step 7. the room in the editor will be preserved.

## Actual Result:
In Step 5, the room in the message is not copied.
In Step 7. the room in the editor is missing.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/d46d3c24-9fcd-430f-b603-c81d915f9d8e

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c98cde11a4c25f4e</li>
        <li>Upwork Job ID: 1780934156183453696</li>
        <li>Last Price Increase: 2024-07-10</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by izarutskaya** [comment]
@JmillsExpensify I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by izarutskaya** [comment]
We think this issue might be related to the #collect project.

**Comment by lakchote** [comment]
Not a blocker - it's an edge case and no core customer functionality is broken.

**Comment by Santhosh-Sellavel** [comment]
Roomname is copied successfully, but the editor is still empty while editing the room mention

**Comment by JmillsExpensify** [comment]
Interesting, let's keep this open then

**Comment by ahmedGaber93** [comment]
@Santhosh-Sellavel room name will be copied successfully only with optimistic data like when user offline, but messages with backend data will produce the issue.

https://github.com/Expensify/App/assets/41129870/ebc47231-b354-4e78-802c-0b7af9eccd24

**Comment by Santhosh-Sellavel** [comment]
> but messages with backend data will produce the issue.
That's address by your proposal @ahmedGaber93 

Now here we wait for 
> the editor is still empty while editing the room mention

**Comment by lakchote** [comment]
What's the latest on here @Santhosh-Sellavel @ahmedGaber93?

**Comment by ahmedGaber93** [comment]
@lakchote [My proposal](https://github.com/Expensify/App/issues/40403#issuecomment-2064017157) here should fix copy to clipboard issue, but @Santhosh-Sellavel mention here https://github.com/Expensify/App/issues/40403#issuecomment-2068185779 another case has the same root cause which not covered by my proposal.

The root cause is the backend data come with mention based-reportID format `<mention-report reportID="1234" />`, and `ExpensiMark` in `replace` method not have any rule to replace it with mention based-name `<mention-report>#romName</mention-report>`.

I think it is difficult to make `ExpensiMark` replace mention based-reportID to mention based-name, because we need to the reports data from onyx, and pass it to `ExpensiMark` seems unclean.

So I think we can apply the same fix in my proposal to everywhere we use report mention (BE handled it now in comments only), by create a new method `replaceMentionReport(html)`

```javascript
function replaceMentionReport(html) {
  return html.replace(/<mention-report reportID="(\d*)"\/>/gi, (match, reportID): string => {
     const report = getReport(reportID);
     return `<mention-report>${report?.reportName ?? report?.displayName}<mention-report/>`;
  })
}
```

**Comment by Santhosh-Sellavel** [comment]
> The root cause is the backend data come with mention based-reportID format `<mention-report reportID="1234" />`, and `ExpensiMark` in `replace` method not have any rule to replace it with mention based-name `<mention-report>#romName</mention-report>`.

If so then we should add/update rule there

**Comment by Santhosh-Sellavel** [comment]
@ahmedGaber93 replacing everywhere seems weird too.
How many places should we replace this?  Can you list down?

**Comment by ahmedGaber93** [comment]
After dig in `ExpensiMark.js` I found we already have parameter [`extra`](https://github.com/Expensify/expensify-common/blob/1713f28214f0e7176c4fd13433fb0ea15491ebf9/lib/ExpensiMark.d.ts#L37-L40) which have `accountIDToName` and `reportIdToName` in `htmlToMarkdown` and `htmlToText` and will replace them [here](https://github.com/Expensify/expensify-common/blob/1713f28214f0e7176c4fd13433fb0ea15491ebf9/lib/ExpensiMark.js#L467-L478), but I can't found any usage for this parameter in the App, also when trying to use it, it not works, but I will try again.

**Comment by ahmedGaber93** [comment]
>  How many places should we replace this? Can you list down?

I think two places only [edit action](https://github.com/Expensify/App/blob/334b630c7dc4884b21937d3972820d02ed039704/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L241) and [copy to clipboard](https://github.com/Expensify/App/blob/9b839f4f18d724dba1ed17ed5967f7e033781428/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L398) because BE replace report mention in comments only.

**Comment by ahmedGaber93** [comment]
> when trying to use it, it not works, but I will try again.

@Santhosh-Sellavel After some investigation, I found why it not works (see step 3, 4), and these are the changes that work with me.
1. in src/libs/ReportUtils.ts add new method `getReportIdToName`
```javascript
function getReportIdToName(): Record<string, string> | undefined {
    const accountIDToName: Record<string, string> = {};
    Object.values(allReports ?? {}).forEach((report: OnyxEntry<Report>) => {
        if (report?.reportID && report?.reportName){
            accountIDToName[report?.reportID] = report?.reportName;
        }
    });
    return accountIDToName;
}
```

2. in [setClipboardMessage](https://github.com/Expensify/App/blob/9b839f4f18d724dba1ed17ed5967f7e033781428/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L43-L51) pass `reportIdToName` to `ExpensiMark`
```javascript
function setClipboardMessage(content: string) {
    const parser = new ExpensiMark();
    const reportIdToName = ReportUtils.getReportIdToName();
    if (!Clipboard.canSetHtml()) {
        Clipboard.setString(parser.htmlToMarkdown(content, {reportIdToName}));
    } else {
        const plainText = parser.htmlToText(content, {reportIdToName}); // this now return #roomName
        Clipboard.setHtml(content, plainText);
    }
}
```

3. in `setClipboardMessage` above we fix the plainText but when we copy inside composer in web we check for html (content) first if found we copy it not plainText see [here](https://github.com/Expensify/App/blob/d0513b9513a091cb49f79fbc429b73d2a987cfce/src/hooks/useHtmlPaste/index.ts#L107).
```javascript
  Clipboard.setHtml(content, plainText);
  // we now copy this values Clipboard.setHtml('<mention-report reportID="1234" />', '#roomName')
  
  // paste in composer will paste content `<mention-report reportID="1234" />` return empty text and still need to fix
  // past outside app will past plainText `#roomName` fixed
```

To fix `content` above, we need to pass reportIdToName in [handlePastedHTML](https://github.com/Expensify/App/blob/d0513b9513a091cb49f79fbc429b73d2a987cfce/src/hooks/useHtmlPaste/index.ts#L66) to be `parser.htmlToMarkdown(html, {reportIdToName})`.

4. we still have a problem when I copy `<mention-report reportID="1234" />` as HTML to clipboard, the paste value [here](https://github.com/Expensify/App/blob/d0513b9513a091cb49f79fbc429b73d2a987cfce/src/hooks/useHtmlPaste/index.ts#L109) paste with different format `<mention-report reportid="1234"></mention-report>` (`<openTag></closeTag>` not `<closedTag />`), and this format not match rule `reportMentions` in `ExpensiMark` so step 3 above will failed.

So we need to edit `reportMentions` [here](https://github.com/Expensify/expensify-common/blob/1713f28214f0e7176c4fd13433fb0ea15491ebf9/lib/ExpensiMark.js#L469) to match `<closedTag />` and `<openTag></closeTag>` 
```
// regex: /<mention-report reportID="(\d+)" *\/>/gi,
regex: /<mention-report reportID="(\d+)"(?: *\/>|><\/mention-report>)/gi,
```

https://github.com/Expensify/App/assets/41129870/78f37c5f-9cbd-43a3-8d61-9288f51cc594

**Comment by lakchote** [comment]
@Santhosh-Sellavel can you take a look at recent comments from @ahmedGaber93 please?

**Comment by eh2077** [comment]
@lakchote We can put this on hold as discussed here https://github.com/Expensify/App/issues/40477#issuecomment-2093804873. We can have a holistic solution to fix them together.

**Comment by marcaaron** [comment]
@ahmedGaber93 do you think you could drop a sort of Problem & Solution statement that summarizes both of these very similar issues and present the holistic plan to fix them?

**Comment by ahmedGaber93** [comment]
> @ahmedGaber93 do you think you could drop into https://github.com/Expensify/App/issues/40477#issuecomment-2093804873 a Problem & Solution statement that summarizes both of these very similar issues and present the holistic plan to fix them?

@marcaaron Yes I agree, these are two implementations for one feature, and fixing them together will facilitate the process of testing and discover the possible bugs and fix it.

**Comment by eh2077** [comment]
I added a summary here https://github.com/Expensify/App/issues/40477#issuecomment-2104490261

**Comment by eh2077** [comment]
We have updates on the other issue.

**Comment by JmillsExpensify** [comment]
Thanks, yes updated on the other issue.

**Comment by JmillsExpensify** [comment]
Based on the above discussion, should we close this issue or leave it open?

**Comment by lakchote** [comment]
> Based on the above discussion, should we close this issue or leave it open?

Looking at latest @eh2077's [comment](https://github.com/Expensify/App/issues/40477#issuecomment-2113052963), we could put this on hold for https://github.com/Expensify/App/pull/40565 to keep track of the issue, and revisit once it's merged?

**Comment by JmillsExpensify** [comment]
Fine putting it on Hold for now. That PR is being actively worked on.

**Comment by JmillsExpensify** [comment]
Moved to weekly given the hold

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Looks like the related issue is now merged, so removing the hold.

**Comment by lakchote** [comment]
@eh2077 the related PR https://github.com/Expensify/App/pull/40565 that we were waiting for was merged, but it did not solve the issue.

**Comment by eh2077** [comment]
@lakchote Yes, I pinged @ahmedGaber93 here https://github.com/Expensify/App/issues/40477#issuecomment-2199950787. I think it still makes sense to solve it together with https://github.com/Expensify/App/issues/40477

**Comment by eh2077** [comment]
We're fixing this together with https://github.com/Expensify/App/issues/40477 - waiting for the PR from @ahmedGaber93

**Comment by JmillsExpensify** [comment]
Looks like we're waiting on the related App PR, though I'll go ahead and keep this issue as is because it shouldn't be too much longer.

**Comment by mallenexpensify** [comment]
Closing, we'll handle payment and regression tests in the other issue
- https://github.com/Expensify/App/issues/40477#issuecomment-2243996963
